src/charm.py

```diff
@@ -214,13 +216,15 @@ def _push_alert_rules(self, alerts):
         """
         # Without multitenancy, the default is `anonymous`, and the ruler checks under
         # {RULES_DIR}/<tenant_id>
-        tenant_dir = f"{RULES_DIR}/anonymous"
+        tenant_dir = f"{RULER_STORAGE_DIR}/anonymous"
+        # Need to `make_dir` even though we have `make_dirs=True` below
+        # https://github.com/canonical/operator/issues/898
         self._container.make_dir(tenant_dir, make_parents=True)
         for topology_identifier, rules_file in alerts.items():
             filename = f"juju_{topology_identifier}.rules"
             path = os.path.join(tenant_dir, filename)
             rules = yaml.safe_dump(rules_file)
-            self._container.push(path, rules, make_dirs=True)
+            self._push(path, rules)
```

Comment:
Let's **not** make this into a common antipattern. 

In addition, we don't need the hash. We also don't need to restart Mimir if they change. It does not have the same behavior as Prometheus. Let's drop that behavior from Mimir while we're at it (it doesn't need to be in the restart check, we don't need a hash, and that also means we don't need `_push|_pull`).

When we get to a "distributed" rules store (I could say "if we do", but if we want to support users adding their own rules in a non-Juju way through the API, we will need to), the logic involved will need to drive `mimirtool` and/or the Mimir API directly.

Until we get to that point, when `"local"` is used, every time the timer for the ruler runs, it invokes [this function](https://github.com/grafana/mimir/blob/cc04dbd807f49f116815cf13a4af4ea3e50eedf6/pkg/ruler/rulestore/local/local.go#L83), and that invokes non-exported functions in the same file which [read the path](https://github.com/grafana/mimir/blob/cc04dbd807f49f116815cf13a4af4ea3e50eedf6/pkg/ruler/rulestore/local/local.go#L117), from the beginning. Every time we ask the API endpoint, it will do this. 

There's no need to hash, because we don't need to restart Mimir, even for the integration tests. Put them in the folder, query the endpoint, and they're there. That means there's also no need for `_push|_pull`

Reply:
> every time the timer for the ruler runs, it invokes [this function](https://github.com/grafana/mimir/blob/cc04dbd807f49f116815cf13a4af4ea3e50eedf6/pkg/ruler/rulestore/local/local.go#L83), and that invokes non-exported functions in the same file which [read the path](https://github.com/grafana/mimir/blob/cc04dbd807f49f116815cf13a4af4ea3e50eedf6/pkg/ruler/rulestore/local/local.go#L117), from the beginning. Every time we ask the API endpoint, it will do this.

Need to be careful of implementation leakage.
Iiuc, you're saying that:
1. on_remote_write_relation_changed should also make an api call?
2. we need an itest in which we update the rules and look at `/alerts` without looking at `/rules` first, and make sure the alert still fires?

Reply:
I'm saying that, at some point in the future, we will need to make an API call. Not also a call. Replacing this. API management of rules and local folder scanning cannot be used together, and if we must choose whether to tell users they cannot do programmatic management (which Grafana's alerting API can also do) or scan, it's API. 

Also because scaling Mimir out and distributing evaluation must be done that way. 

We do not need that integration test. We need one which asserts that the rules are present/absent/etc _without_ restarting Mimir. Whether the alert is actually firing isn't that crucial